Update the API class to properly put the access_token into the header.

anton-molyboha opened the pull request:
This pull request:

- Creates an `Api_new` class that mirrors the old `Api` class, but
   1. All api functions return a namedtuple `HttpRequest`, that includes headers information.
   2. All authorization is done via "Authorization" header rather than the query parameter.
- Migrate AsyncClient and HttpClient to use the API_new class.

This should be a non-breaking change at this point. I have run the existing tests, and they pass, but I did not try to improve test coverage; bugs could slip through the cracks.

The next steps would be to:
1. Rename `Api_new` -> `Api`, while having the old `Api` class available as `Api_old`.
2. Delete `Api_old`.
Those will be braking changes for those, who depend directly on the Api class, but users will have a path to upgrade.

anton-molyboha commented:
Forgot to mention the related discussion: https://github.com/matrix-nio/matrix-nio/issues/511

PaarthShah reviewed (commented):
At a high level, I think we should avoid all `_new` naming conventions, both for classes and functions. The `_old` pattern is fine, but I'd make a point of using the `warning` stdlib to introduce a deprecation warning.

There are a few reasons for this:

1. It's ugly and bloats the number of changes
2. Anyone downstream that may have overridden these classes/names is in for a bad time. Leaving the names the same means there's a good chance that everything will continue to work as long as they invoke `super()`. If they do depend on existing behavior, then let them pin an older version, or update their code.
3. Considering how long the matrix spec has supported the use of these headers, I don't have any serious concerns that we need to worry about backwards compatibility with non-compliant servers
4. By not being at its 1.0 release, `matrix-nio` *itself* (by way of semver) isn't making any promises of being perfectly backwards compatible, and for something this low level, I presume nobody  would actually notice this difference in behavior (this relates to both 3 and 4).

anton-molyboha commented:
> It's ugly and bloats the number of changes. I can't review this properly in this state, but I'll be willing if it goes down to half the size or less.

Fair point. I have removed "Api_old" and renamed "Api_new" back to "Api". Now the combined diff is "only" ~600 lines -- and looking through it did help me catch several details that I have missed in my original PR.

The downside: I'm afraid migrating to the new `Api` became harder. Previously one could migrate part of their code at a time, running tests in between. Now you have to migrate everything at once.

(There was at least one person who mentioned in the matrix chat that they were using the `Api` class directly. That was about half a year ago, though)
